### PR TITLE
WASI Preview 2: rewrite streams and pollable implementation

### PR DESCRIPTION
**Pull request by pchickey**
When we landed the WASI Preview 2 support in `wasmtime-wasi` we knew the streams/poll/scheduler implementation was going to get torn out and rewritten: see task 4 in https://github.com/bytecodealliance/wasmtime/issues/6370. This is that rewrite. Most of this was pair-programmed with @elliottt.

# Features 
`wasmtime-wasi` Preview 2 support is now built on top of `tokio`, and the host implementations for streams and pollable always use `async` Rust. It now passes all of the poll_oneoff tests, including on windows.

The new `HostInputStream`, `HostOutputStream`, and `HostPollable` traits are used to implement the `wasi:io/streams.{input-stream,output-stream}` and `wasi:poll/poll.pollable` resources.

## `Host{Input, Output}Stream` design

`input-stream` and `output-stream` resources have a significant number of methods. For the host traits, we managed to boil this down to two required methods, as well as some optional methods to cover vectored IO, splice, and write-zeroes. The default implementations provide an unoptimized but correct implementation - the user may override them if an optimization is desired.

`HostInputStream` has just two methods: `fn read(&mut self, ...)` and `async fn ready(&mut self)`. The most important invariant is that `read` must be non-blocking, and `ready` must block until the stream is ready for reading.

## Nonblocking IO is actually impossible on regular files, but we hide that detail from WASI programs

There's a huge lie in the description above: internally, wasmtime_wasi uses `InternalHost{Input,Output}Stream`, which are enums that contain either a `Host{Input,Output}Stream`, or a private `File{Input,Output}Stream`. These file streams always perform a blocking read/write, whether or not the WASI program has called the nonblocking `read` or the blocking `blocking_read` method. And, any pollable associated with file read or write readiness always returns ready immediately.

This is required because there is no way to implement streams on files with `epoll(7)` (or on macos or windows using mio, afaik) based readiness - `poll(2)` will lie and always tell you files are ready for reading of writing, but the actual syscall may end up blocking the thread due to disk IO. And, rather than stick the work on a background task, we must wait until the OS `read(2)` or `write(2)` for a file has completed before returning from wasi stream (nonblocking) read or write, because the userlands in WASI expect (reasonably!) that any errors in regular file IO are reported synchronously.

So, we put an escape hatch into wasmtime-wasi which is just for regular file IO. It `awaits` for a tokio `spawn_blocking` to complete in both the blocking and non-blocking implementations of the stream reads and writes. Like all the other syscalls in the filesystem implementation, we need the `async` rust to put any blocking syscalls in a `spawn_blocking`, in order to not block the executor.

We spent a the better part of a week trying to square this circle and this is the best we could come up to make files work like we expect, but also expose the right interfaces from wasmtime-wasi so that all other streams have the proper non-blocking semantics.

## `HostPollable` design

The trick to `HostPollable` is to create a Rust `Future` indicating whether a stream (or some other resource, e.g. http bodies) is ready without requiring both the stream and pollable hold some `Arc<Mutex<...>>` to the resource itself. @alexcrichton pointed out the right design here: `HostPollable` is an enum with two variants:
*  `TableEntry { index: u32, make_future: for<'a> fn(&'a mut dyn Any) -> Pin<Box<dyn Future = Result<()>> + Send + 'a>}` means, get a mutable reference to some (still dynamically-typed) other entry in the table, and apply this fn to create a Future from that entry, which has the same lifetime as the mutable reference. For the streams, we use this variant, and make_future is just a call to the `Host{Input,Output}Stream::ready` method.
* `Closure( Box<Fn() -> Pin<Box<dyn Future = Result<()>> + Send + 'static>> + Send + Sync + 'static> )` means, use this owned closure to create a Future with a static lifetime. This is used for creating timers from the monotonic and wall clocks, which are ambient and therefore do not have table entries.

Note that we discovered a major design problem with the `TableEntry` variant that exposes a crash vector if the parent resource of a pollable is destroyed before the pollable itself. This is a big enough problem that we are setting it aside to be solved in a follow-up PR.

## Implementations

`wasmtime-wasi` provides a struct `AsyncReadStream` which wraps a `tokio::io::AsyncRead` to provide an impl of `HostInputStream`, `AsyncWriteStream` to wraps `AsyncWrite` to provide `HostOutputStream`.

`AsyncReadStream` and `AsyncWriteStream` will manage their own buffer in order for the `HostInputStream::read` and `HostOutputStream::write` methods to be non-blocking.  Each spawns a helper tokio::task to perform async operations in the background. This requires some buffering.

Note, we discussed ensuring these background tasks are reaped properly - right now we believe that the implementations are correct and they will exit on their own when the foreground side gets dropped. However, in the future to implement e.g. splice and forward correctly, we may need to allow these tasks to (optionally!) live longer than the WASI program's execution.

Additionally, we have provided an implementation that correctly handles waiting on `stdin` readiness on both windows and unix. In Unix, this is built on top of `tokio::os::unix::AsyncFd`, allows tokio to register stdin with `epoll(7)`. Because stdin is a process-wide singleton, and also because epoll will return an error if the same fd is registered multiple times, the stdin resource is a global singleton, created lazily and guarded by a mutex. `wasmtime_wasi::preview2::stdio::stdin()` returns a `struct Stdin` which takes a lock on that singleton in the `HostInputStream::{read, ready}` methods. This means that, if for some reason you have granted the process stdin to multiple wasi streams (in the same, or different, stores) you can get all sorts of strange behavior because the same global resource backs them all behind the scenes - but this is always the case, the implementation in wasmtime_wasi just allows you to do so without hanging or panicking.

On Windows, the implementation of stdin is broken, and inheriting stdin will panic. We have a path to fixing it, but we are leaving that out of this PR, because it has been open for far too long, and far too much work is predicated on landing it.

# Sync and Async uses

Despite using async for its implementation, `wasmtime-wasi` will still work for synchronous Wasmtime embeddings. A new `test-programs` suite `wasi-preview2-components-sync` is identical to `wasi-preview2-components` in all regards except that it uses wasmtime's synchronous APIs, and the appropriate `wasmtime_wasi::command::sync::add_to_linker`, rather than `wasmtime_wasi::command::add_to_linker` which requires an async Config. The synchronous implementation creates its own Tokio runtime (if one does not already exist) and then invokes the async implementation inside of a `block_on`.

The `wasi-common` Preview 1 `poll_oneoff` implementation allowed for a synchronous implementation to work without any sort of dependency on tokio, by abstracting it behind a pluggable `WasiSched`. This PR eliminates that option for our Preview 2 users. If there is a compelling need, and labor available to implement and maintain it, we could collaborate on re-introducing a fully synchronous tokio-free option in the future. For this implementation, I decided that we just need a tokio backend in order to build a path towards supporting both wasi-sockets and wasi-http with the same modular stream and pollable host traits.

# Bindings traits

After living with the bindings generated underneath `crate::preview2::wasi` and `crate::preview2::wasi::command` for a while, I decided that the interface traits belong under `crate::preview2::bindings` and command belongs under `crate::preview2::command`.

Beyond that code motion, we now generate async traits for the `wasi:io/streams`, `wasi:poll/poll`, and `wasi:filesystem/filesystem` interfaces, and sync traits for everything else. If you want to use the bindings from an async embedding (i.e. with an async Config), the `crate::preview2::bindings::{interface}::add_to_linker` will do it, and if you want to use them from a synchronous embedding, use `crate::preview2::bindings::sync::{interface}::add_to_linker`.

### CODE REVIEW

File: crates/wasi/src/preview2/poll.rs
Comment (dicej) [reply crates/wasi/src/preview2/poll.rs:85]: I should mention that we wouldn't get the full benefit of `FuturesUnordered` here given the one-off nature of `poll-oneoff`, so it might not be worth it until/unless we replace `poll-oneoff` with a resource that's intended to be polled repeatedly.

Comment (alexcrichton) [reply crates/wasi/src/preview2/poll.rs:85]: I definitely don't disagree about the benefits of `FuturesUnordered`, but I would personally say we should start off simpler here and see how things go. My guess is that a large number of futures would have many other inefficiencies other than the polling behavior here, and in the meantime, at least to me personally, I feel that a manual "poll the list" is easier to understand than the `FuturesUnordered` abstraction

File: crates/wasi/src/preview2/stream.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/stream.rs:253]: This doesn't feel quite right to me in that this shouldn't continuously ressize the buffer but instead if the buffer is non-empty then it's always ready for reading but a read is only done if the buffer is empty (in which case we could for example know it's already been initialized)

File: crates/wasi/src/preview2/stream.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/stream.rs:251]: This is a bit tricky because the futures returned from `ready` need to all be "cancel safe", so if there's pending data in this buffer then it'll get lost if the `await` point below is cancelled.

File: crates/wasi/src/preview2/stream.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/stream.rs:295]: Sort of like above with pipes but I think this'll want a form of backpressure to avoid the internal buffer from growing too large

File: crates/wasi/src/preview2/stream.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/stream.rs:307]: The error here is tricky to handle because data has been buffered up from the beginning of the function but this is also reporting an error, so if the application re-wrote the data then it'd be written twice

File: crates/wasi/src/preview2/stream.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/stream.rs:318]: In terms of "cancel-safety" this needs to be aware of partial writes that succeed, so I don't think this will be able to use `write_all` but instead will have to `loop` and incrementally update the internal buffer.

**Review by alexcrichton — COMMENTED** [review]
Looking reasonable so far, but the high level things which I think need more looking into are:

* The `block_in_place` place function I don't think can ever be used correctly since it's used in contexts that shouldn't block for the express purpose of blocking.
* Buffering internally I think needs to all be "capped" so it doesn't buffer up too much data on repeated non-blocking calls to write/read
* I only realized this later in review, but all `ready` futures need to be "cancel safe" where the internal objects are all left in a consistent state if the operation is cancelled.

**Review by dicej — COMMENTED** [review]
Looks like a great start overall.

Echoing Alex's concerns about `block_in_place`:
- We should never use it in any function which must be non-blocking from the guest's perspective.
- We should probably also not use it in any function which _is_ blocking from the guest's perspective since it blocks the Wasmtime fiber, silently inhibits the composition of futures (e.g. for timeouts), and also blocks the embedder in a way it might not expect.

So we can use `spawn_blocking` instead, and store the future as part of the stream, but at that point we're kind of reinventing `tokio::fs`.  The `tokio` folks have already provided a battle-tested implementation, and `tokio::fs::File` implements `AsyncRead` and `AsyncWrite`, so I think we should reuse it wherever possible.  I realize there are some cases where we need to use `cap-std` to enforce sandboxing, but once `cap-std` gives us a file handle, we should be able to convert it to a `tokio::fs::File` and use that to handle the details of bridging the async and sync worlds.  Does that sounds plausible?

Regarding buffers: I don't think we should ever need to grow buffers beyond their original sizes (and certainly not without bound).  I would expect that we would flush (for writes) or fill (for reads) a buffer before moving on to the next one (or reusing it without necessarily changing its size).  If you believe there's a scenario where extending a buffer is unavoidable (or at least desirable), I'd like to have a conversation to understand that better.

Generally speaking, I would expect that neither `HostInputStream::read` nor `HostOutputStream::write` implementations would do _any_ I/O and would instead leave that to `ready`, which would fill or drain `self.buffer`, respectively.  That would avoid the need for the no-op waker and any kind of `block_in_place` trickery.  Or put another way, I/O should only be done in an async context where a real waker is available, thus guaranteeing wake events make it back to `poll_oneoff`.

Hope that's somewhat clear.  Happy to discuss in more detail as necessary.

**Comment by alexcrichton** [comment]
Perhaps a point in favor of using tokio-native types is that it might provide a nicer error and/or solution to the panic outlined at https://github.com/bytecodealliance/wasmtime/issues/5306#issuecomment-1625917098

File: crates/wasi/src/preview2/pipe.rs
Comment (pchickey) [reply crates/wasi/src/preview2/pipe.rs]: Yeah, we'd have to do some challenging stuff to destroy the ctx, which is harder because the methods need both the ctx and table via WasiView, so we decided against changing this now.

We really only expect this impl to be used for tests.

File: crates/wasi/src/preview2/poll.rs
Comment (pchickey) [reply crates/wasi/src/preview2/poll.rs:109]: An io error that occurs is buffered up, but we wanted to leave it open for `ready()` to trap execution if something is horribly wrong.

File: crates/wasi/src/preview2/preview2/io.rs
Comment (pchickey) [reply crates/wasi/src/preview2/preview2/io.rs:241]: Yes this is a pretty significant complication. Right now we have decided to punt on it by adding a big warning in the source code, land this PR, then see if any of the assumptions involved change as we switch over to resources, and then fix it. It would be a crash vector for the whole runtime, but it would also require some pretty major work, so I want to defer it to a future PR. We will fix it before we mark the preview2 implementation as production ready.

File: crates/wasi/src/preview2/pipe.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/pipe.rs:53]: I think that this may always want to be `Ok(())` since `read` never blocks?

File: crates/wasi/src/preview2/ctx.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/ctx.rs:39]: Given the semantics of that this stream is never ready for reading, would a better default be the `ClosedInputStream`?

File: crates/wasi/src/preview2/pipe.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/pipe.rs:360]: I find the semantics of this type a bit dubious, but is this necessary? Would it be possible to remove this?

File: crates/wasi/src/preview2/pipe.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/pipe.rs:281]: Should this case be `unreachable!` since if the state is "ready" then the task should always be trying to receive one buffer?

File: crates/wasi/src/preview2/pipe.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/pipe.rs:217]: Instead of an `Option` here perhaps add `Closed` to `WriteState`?

File: crates/wasi/src/preview2/preview2/filesystem.rs
Comment (alexcrichton) [inline crates/wasi/src/preview2/preview2/filesystem.rs:606]: Had a good chuckle reading this :)

Regardless I agree we can't make this truly async and sync close here is the right thing to do IMO

**Review by alexcrichton — APPROVED** [review]
Nice 👍 

For the WIT changes, those are from the upstream proposals? Or local changes made temporarily for this PR?

File: crates/wasi/src/preview2/pipe.rs
Comment (pchickey) [reply crates/wasi/src/preview2/pipe.rs:53]: Correct.

File: crates/wasi/src/preview2/pipe.rs
Comment (pchickey) [reply crates/wasi/src/preview2/pipe.rs:360]: yeah, we did it because it corresponded to one of the output streams but now that you point it out, its really not necessary at all

File: crates/wasi/src/preview2/pipe.rs
Comment (pchickey) [reply crates/wasi/src/preview2/pipe.rs:277]: correct- that was left over from a prior factoring of this code, but its unreachable now

File: crates/wasi/src/preview2/pipe.rs
Comment (pchickey) [reply crates/wasi/src/preview2/pipe.rs:213]: we had it factored that way, then we were taking advantage of Option::take() so changing the repr made sense, but now I think I can switch it back...

**Comment by pchickey** [comment]
The wit has not been submitted upstream yet. I was going to work on that once this landed, as well as the rest of the upstream/wasmtime wit reconciliation that is pending

**Comment by pchickey** [comment]
I just made a big update to the PR description with the design changes since the last round of review, and the known problems in this PR. I am now going to make the windows inherit stdin panic, mark the tests that affects as cfg_attr(windows, should_panic), and land this PR. The follow up work includes, in rough priority order:

* HostPollable::TableEntry issues around pollables outliving their parent resource, which are a crash vector right now
* get windows stdin working, and better tests of unix asyncfd based stdin
* implementing stream splice and forward, and dealing with the lifetimes of background tasks accordingly

**Pull request merged**
